### PR TITLE
Add Hashmap slot+3keys (3D) macros

### PR DESCRIPTION
**Pull request by goncaloMagalhaes**
* Add Hashmap 3D macros
* Add testing for those macros
* Add missing testing for 2D macros
* Remove 'view' type from store functions definitions in HashmapWrappers

### CODE REVIEW

File: src/data-structures/Hashmap.huff [hunk 1]
```diff
@@ -62,6 +62,47 @@
...
 
+/// @notice Calculate the slot from three keys
+#define macro GET_SLOT_FROM_KEYS_3D(mem_ptr) = takes(4) returns (1) {
+    // Input stack: [slot, key1, key2, key3]
+    // Load the data into memory
+    <mem_ptr>           // [<mem_ptr>, slot, key1, key2, key3]
+    mstore              // [key1, key2, key3]
+
+    // next byte
+    <mem_ptr> 0x20 add  // [<mem_ptr> + 32, key1, key2, key3]
```
Comment (MathisGD) [inline src/data-structures/Hashmap.huff:73]: small opti but you could keep the value in the stack, as it is re-used two times after

File: test/data-structures/mocks/HashmapWrappers.huff [hunk 2]
```diff
@@ -1,18 +1,45 @@
...
+#define function loadElementFromKeys2D(bytes32, bytes32, bytes32) view returns (bytes32)
+
+// Returns the slot value by hashing the given slot and three keys.
+// sig: 0x0cc703f5
+#define function loadElementFromKeys3D(bytes32, bytes32, bytes32, bytes32) view returns (bytes32)
+
 // Stores the value for the given key.
 // sig: 0x376caf9f
-#define function storeElement(bytes32 key, bytes32 value) view returns ()
+#define function storeElement(bytes32 key, bytes32 value) nonpayable returns ()
```
[the remark this answers is not in the document]
Comment (goncaloMagalhaes) [reply test/data-structures/mocks/HashmapWrappers.huff:19]: Doesn't need to, though it's not supposed to be `view` and putting it as payable seems misleading to me.

File: src/data-structures/Hashmap.huff (hunk 1 above)
Comment (goncaloMagalhaes) [reply src/data-structures/Hashmap.huff:73]: Not sure I understood that. How would the code look like then?

Comment (MathisGD) [reply src/data-structures/Hashmap.huff:73]: You would dup `<mem_ptr> + 32` to be able to reuse it after with just a dup

File: test/data-structures/mocks/HashmapWrappers.huff (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/data-structures/mocks/HashmapWrappers.huff:19]: Isn't it possible to put nothing (neither view, payable or nonpayable) ?

File: src/data-structures/Hashmap.huff (hunk 1 above)
Comment (goncaloMagalhaes) [reply src/data-structures/Hashmap.huff:73]: I believe I followed the same pattern as the other Hashmap.huff functions. Maintainers can chime in on whether that optimization is necessary/wanted.

Comment (cheethas) [reply src/data-structures/Hashmap.huff:73]: now there are three uses keeping it on the stack makes sense! With just two a dup1 and a swap are as expensive as another push (usually) so it is ignored.

File: test/data-structures/mocks/HashmapWrappers.huff (hunk 2 above)
[the remark this answers is not in the document]
Comment (cheethas) [reply test/data-structures/mocks/HashmapWrappers.huff:19]: unfortunately one is needed by the compiler, probably would make sense to remove that requirement in the future

**Review by cheethas — COMMENTED** [review]
Looks good, I think we should have a go at implementing the optimisation Mathias has suggested!

**Review by cheethas — APPROVED** [review]
based

**Pull request merged**